### ISSUE TITLE
Keras network works in some machines, but predicts same class in others

### ISSUE THREAD

**Issue by kilesw** [keras-team/keras#12562]
Please make sure that the boxes below are checked before you submit your issue.
If your issue is an **implementation question**, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [on the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) instead of opening a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
`pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps`

- [x] Check that your version of TensorFlow is up-to-date. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

I used a pretrained InceptionV3 network with imagenet weights to transfer learning to an image classification problem with 14 classes. The network reaches a 92 % accuracy while training, and the first time I used it (with CPU) it achieved an 86 % over the test set. I moved to GPU after that, and since that moment the network started predicting most of the test images with the same class. The training process still achieves a 92 %. Predictions don't work even if I move to CPU again. I have tested the code in other machines, and it works only in some of them, independently of the use of GPU or not. Here is my code:

    from keras.preprocessing.image import ImageDataGenerator,array_to_img, img_to_array, 
    load_img
    from keras.layers import Dense
    from keras.models import Model
    from keras.applications.inception_v3 import InceptionV3
    import numpy as np
    import pandas as pd
    from PIL import Image
    import os

    # Fix seed
    seed = 1234

    # Load test data
    generator = ImageDataGenerator()
    iterator = generator.flow_from_directory('Test_Mixed',target_size=(299, 299),batch_size=146, 
    shuffle=False)
    test, y = iterator.next()
    names = iterator.filenames
    names = [ name[name.index('/')+1:] for name in names]

    # Load pre-trained model 
    base_model = InceptionV3(weights='imagenet', include_top=False, pooling = 'avg',input_shape=(299, 299, 3))

    for layer in base_model.layers:
        layer.trainable = False
 
    output_layer = base_model.output

    # Hidden and output layers to train
    hidden = Dense(100,input_dim=1000,activation='relu')(output_layer)
    final = Dense(14,input_dim=32,activation='softmax')(hidden)

    # Final model
    model = Model(inputs=base_model.input,outputs=final)
    model.compile(optimizer='rmsprop',loss='categorical_crossentropy',metrics=['accuracy'])

    # Data augmentation
    generator = ImageDataGenerator(
                    #rescale=0.7,
                    rotation_range=20,
                    width_shift_range=0.2,
                    height_shift_range=0.2,
                    horizontal_flip=True
                )

    iterator = generator.flow_from_directory('train/',target_size=(299, 299), batch_size=32, 
    shuffle=True, seed = seed)

    # Training
    model.fit_generator(iterator,epochs=15,steps_per_epoch=20,verbose=True)

    # Predictions
    y_preds = model.predict(test)
    labels = y_preds.argmax(axis=1)

The `model.predict()` function is what outputs the unexpected results. As I have said, during the training process in `fit_generator` the accuracy increases always very nicely, but then predictions are bad. To test overfitting I have also tried to use `model.predict()` with the training set, obtaining bad results again. What I find more confusing is the dramatic changes from  other computers (and the first attempt in my machine) to the results happening now: from 86 % test accuracy to same class predictions, which is less than 10 % accuracy. The code has been the same all this time.

**Comment by ariyabolouri** [comment]
It appears you're describing the classic overfitting problem, whereby the model learns how to increase its accuracy on the training set "too well". Instead of a colour detector being able to identify pictures of grass are green or a sky is blue, an over-trained model might learn to detect patterns within the grass that might not be reproducible when testing against a picture of a meadow, for example. 

This problem can be mitigated by adding a noise layer to randomly reduce some of the input data, so that the model will have to learn how to make sweeping generalisations rather than abnormally specific rules. 

https://keras.io/layers/noise/

**Comment by kilesw** [comment]
I think overfitting is not the problem. When I run the code in a friend's computer it always outputs good predictions. The good or bad predictions depend on the used machine.

**Comment by MuLx10** [comment]
May be that has to do with a random seed. I mean network parameters initializations. You can set up a constant seed

**Comment by kilesw** [comment]
Do you mean fixing the seed in the `kernel_initializer` argument? I have also tested it a few days ago. However the results seem still to be machine-dependent. It doesn't mind if I fix a seed or not, the prediction results don't change drastically within the same machine, only from a machine to another.

**Comment by MuLx10** [comment]
I mean this
```python
import os
os.environ['PYTHONHASHSEED'] = '0'

import keras
import random as rn
import numpy as np

np.random.seed(1)
rn.seed(1)

```

**Comment by kilesw** [comment]
Anyways, without setting the seed the differences are not significative in the same machine. I've just tried that and the predictions still are good only in some machines, and same-class in my computer.
